pkg/modules/services/implservices/helpers.go

```diff
@@ -0,0 +1,136 @@
+	variables := make(map[string]qbtypes.VariableItem)
+	parts := make([]string, 0, len(tags))
+	valueItr := 1
+	for _, t := range tags {
+		valueIdentifier := fmt.Sprintf("%d", valueItr)
+
+		switch strings.ToLower(t.Operator) {
+		case "in":
+			if vals, ok := pickInValuesFromTag(t); ok {
+				variables[valueIdentifier] = qbtypes.VariableItem{Type: qbtypes.DynamicVariableType, Value: vals}
+			} else {
+				continue
+			}
+			parts = append(parts, fmt.Sprintf("%s IN $%s", t.Key, valueIdentifier))
+		case "equal", "=":
+			if v, ok := pickEqualValueFromTag(t); ok {
+				variables[valueIdentifier] = qbtypes.VariableItem{Type: qbtypes.DynamicVariableType, Value: v}
+			} else {
+				continue
+			}
+			parts = append(parts, fmt.Sprintf("%s = $%s", t.Key, valueIdentifier))
+		default:
+			// skip unsupported for now
+			continue
+		}
+
+		valueItr++
+	}
+
+	// now also add the condition representing top level operations permitted only
+	filterExpr := strings.Join(parts, " AND ")
+	scopeExpr := fmt.Sprintf("isRoot = $%d OR isEntryPoint = $%d", valueItr, valueItr+1)
+	if filterExpr != "" {
+		filterExpr = "(" + filterExpr + ") AND (" + scopeExpr + ")"
+	} else {
+		filterExpr = scopeExpr
+	}
+	variables[fmt.Sprintf("%d", valueItr)] = qbtypes.VariableItem{
+		Type:  qbtypes.DynamicVariableType,
+		Value: true,
+	}
+	variables[fmt.Sprintf("%d", valueItr+1)] = qbtypes.VariableItem{
+		Type:  qbtypes.DynamicVariableType,
+		Value: true,
+	}
+	return filterExpr, variables
+}
+
+// pickInValuesFromTag returns a []any for IN operator in the precedence order of
+// StringValues, BoolValues, NumberValues. Returns false if none are populated.
+func pickInValuesFromTag(t servicetypesv1.TagFilterItem) ([]any, bool) {
+	if len(t.StringValues) > 0 {
+		vals := make([]any, 0, len(t.StringValues))
+		for _, v := range t.StringValues {
+			vals = append(vals, v)
+		}
+		return vals, true
+	}
+	if len(t.BoolValues) > 0 {
+		vals := make([]any, 0, len(t.BoolValues))
+		for _, v := range t.BoolValues {
+			vals = append(vals, v)
+		}
+		return vals, true
+	}
+	if len(t.NumberValues) > 0 {
+		vals := make([]any, 0, len(t.NumberValues))
+		for _, v := range t.NumberValues {
+			vals = append(vals, v)
+		}
+		return vals, true
+	}
+	return nil, false
+}
+
+// pickEqualValueFromTag returns a scalar any for EQUAL operator using the
+// precedence order string, bool, number. Returns false if none are populated.
+func pickEqualValueFromTag(t servicetypesv1.TagFilterItem) (any, bool) {
+	if len(t.StringValues) > 0 {
+		return t.StringValues[0], true
+	}
+	if len(t.BoolValues) > 0 {
+		return t.BoolValues[0], true
+	}
+	if len(t.NumberValues) > 0 {
+		return t.NumberValues[0], true
+	}
+	return nil, false
+}
+
+// toFloat safely converts a cell value to float64, returning 0 on type mismatch.
+func toFloat(row []any, idx int) float64 {
+	if idx < 0 || idx >= len(row) || row[idx] == nil {
+		return 0
+	}
+	return float64(row[idx].(float64))
```

Comment:
`row[idx].(float64)` this will panic if the `row[idx]` is not a float64 type. it should be handled properly.